containerized build tools [V2]

phlogistonjohn opened the pull request:
After a series of discussions, particularly with Zack C and Dan Mick, I've created a new patch series to replace those in #56711.

This version simplifies the (Container|Docker)file and and replaces the harder to maintain shell script wrapper with a more logical and well constructed python script. Right now there are also a couple of patches present in the branch that I needed to make it compile the unit tests. I plan on moving these out eventually as separate PRs.


PRs and issues related to building in containers:
* <s>https://github.com/ceph/ceph/pull/61030</s>
* <s>https://github.com/ceph/ceph/pull/61031</s>
* https://tracker.ceph.com/issues/68552 - pr https://github.com/ceph/ceph/pull/61327
* https://tracker.ceph.com/issues/69240 - pr https://github.com/ceph/ceph/pull/61326
* <s>https://tracker.ceph.com/issues/69241</s> fixed on main in #61107
* https://tracker.ceph.com/issues/62804
* https://tracker.ceph.com/issues/69525
* https://github.com/ceph/ceph/pull/61354

### How to test & try

1. Get the patches
   Assuming you have a git checkout of ceph/ceph already:
   1. `git remote add -t jjm-containerized-build-pyalt phlogistonjohn git@github.com:phlogistonjohn/ceph.git`
   2. `git fetch phlogistonjohn`
   3. `git switch jjm-containerized-build-pyalt`
   4. (AND/OR) `git switch -c mybranch jjm-containerized-build-pyalt`
5. View the instructions at the top of ./src/script/build-with-container.py
6. Give it a go. Some examples:
   * python3 src/script/build-with-container.py -d centos9 -b build.c9
   * python3 src/script/build-with-container.py -d ubuntu22.04 -b build.u22 -e build -e tests
   * python3 src/script/build-with-container.py -d centos9 -b build.c9 -e rpm 


## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

jmundack commented on src/script/buildcontainer-setup.sh:
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
can we just use the steps at https://docs.ceph.com/en/squid/start/beginners-guide/ to setup the environment? 

as in use `install-deps.sh` so that we can keep all the dependencies per OS etc are all in the same place?

phlogistonjohn replied:
There's a maze of twisty passages throughout these build scripts. This eventually does end up calling install-deps.sh. Without doing a deep check the flow is "include run-make.sh" -> prepare -> install-deps.sh.

I'd love to simplify some of this but this script was added to (a) match more closely what the jenkins builds are doing (b) add extra packages needed to "fix" the container environment where install-deps either misses or doesn't need something but the containerized env does.

jmundack replied:
gotcha.

so is the goal - to deprecate `install-deps.sh` once we get this simplified version done? so that how folks build it locally and how its build in jenkins is consistent(?)

As for the "additional" dependencies needed just for container builds - would it make sense to keep that as a separate list/script so that we can dig into why that difference exists and document it out?

bill-scales replied:
While I'm not a fan of install-deps.sh, I think it is correct that we use this to install all the dependency packages needed for building ceph. I'm not sure about the extra couple of packages that are being installed for the different distro's below - I think these are really missing dependencies and that we should be fixing install-deps.sh so that it installs these packages as well.

jmundack commented on src/script/buildcontainer-setup.sh:
```diff
@@ -0,0 +1,39 @@
+        dnf_clean
+    ;;
+    *~*ubuntu*2?.04)
+        apt-get update
```
is there any reason to run update here? and not use the packages in the base image?

phlogistonjohn replied:
`apt-get update` is needed to download the repo metadata prior to running `apt-get install`. at least that's what I think was the case... but my debian is a bit rusty.

jmundack replied:
my concern on this is 2 fold
1. we will be (afaik) updating all the packages to the latest available one (do we want this to happen every time? i would think we rely on the base ubuntu image being updated?)
2. we dont do this for others - regardless of what we pick - we should do it consistently across the various OS/images (ie run `update` OR just use the latest base image)

phlogistonjohn replied:
> my concern on this is 2 fold
> 
>     1. we will be (afaik) updating all the packages to the latest available one (do we want this to happen every time? i would think we rely on the base ubuntu image being updated?)
> 

I am willing to be corrected (as always) if I am wrong but, and I did just check this locally, calling `apt-get update` doesn't update packages (`apt-get upgrade` does), this *only* downloads metadata.
This does cause it to get the latest metadata, and not a snapshot of the metadata at the point the container was originally created, but I don't think that's what you are concerned about right?

As an example, here's a brief example I just ran to assure myself of what I just wrote:
```
[john@edfu ~]$ podman run -it --rm docker.io/library/ubuntu:24.04
root@f358f91cace7:/# apt-get install cowsay
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package cowsay
root@f358f91cace7:/# 
exit

[john@edfu ~]$ podman run -it --rm docker.io/library/ubuntu:24.04
root@7bc0a8502e6c:/# apt-get update
Get:1 http://security.ubuntu.com/ubuntu noble-security InRelease [126 kB]
Get:2 http://archive.ubuntu.com/ubuntu noble InRelease [256 kB]
Get:3 http://archive.ubuntu.com/ubuntu noble-updates InRelease [126 kB]
Get:4 http://archive.ubuntu.com/ubuntu noble-backports InRelease [126 kB]
Get:5 http://security.ubuntu.com/ubuntu noble-security/main amd64 Packages [630 kB]
Get:6 http://security.ubuntu.com/ubuntu noble-security/multiverse amd64 Packages [15.3 kB]
Get:7 http://security.ubuntu.com/ubuntu noble-security/restricted amd64 Packages [609 kB]
Get:8 http://archive.ubuntu.com/ubuntu noble/universe amd64 Packages [19.3 MB]
Get:9 http://security.ubuntu.com/ubuntu noble-security/universe amd64 Packages [728 kB]
Get:10 http://archive.ubuntu.com/ubuntu noble/main amd64 Packages [1808 kB]            
Get:11 http://archive.ubuntu.com/ubuntu noble/restricted amd64 Packages [117 kB]
Get:12 http://archive.ubuntu.com/ubuntu noble/multiverse amd64 Packages [331 kB]
Get:13 http://archive.ubuntu.com/ubuntu noble-updates/universe amd64 Packages [944 kB]
Get:14 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 Packages [891 kB]
Get:15 http://archive.ubuntu.com/ubuntu noble-updates/multiverse amd64 Packages [19.7 kB]
Get:16 http://archive.ubuntu.com/ubuntu noble-updates/restricted amd64 Packages [679 kB]
Get:17 http://archive.ubuntu.com/ubuntu noble-backports/universe amd64 Packages [11.9 kB]
Fetched 26.7 MB in 2s (12.1 MB/s)                          
Reading package lists... Done
root@7bc0a8502e6c:/# apt-get install cowsay
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libgdbm-compat4t64 libgdbm6t64 libperl5.38t64 libtext-charwidth-perl netbase perl
  perl-modules-5.38
Suggested packages:
  filters cowsay-off gdbm-l10n perl-doc libterm-readline-gnu-perl
  | libterm-readline-perl-perl make libtap-harness-archive-perl
The following NEW packages will be installed:
  cowsay libgdbm-compat4t64 libgdbm6t64 libperl5.38t64 libtext-charwidth-perl netbase
  perl perl-modules-5.38
0 upgraded, 8 newly installed, 0 to remove and 6 not upgraded.
Need to get 8297 kB of archives.
After this operation, 50.6 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

Note that the debian-style package commands do not automatically download the package metadata, unlike `dnf` from the rpm ecosystem. Thus calling `apt-get update` at least once before installing a package is needed. However, it doesn't change the installed packages by itself.

>     2. we dont do this for others - regardless of what we pick - we should do it consistently across the various OS/images (ie run `update` OR just use the latest base image)

I'm not sure what "others" you are referring to here are. Could you please elaborate?

jmundack replied:
by other - i meant `dnf update`

if this is just doing metadata update for the following installs (quirks for `apt-get`) - this makes sense!

(things i learned today! ❤️ )

dmick replied:
@jmundack the dnf equivalent is "dnf makecache", which is often added to your command automatically.  apt makes it separate and willful.

phlogistonjohn commented on src/test/cls_log/test_cls_log.cc:
```diff
@@ -332,7 +332,6 @@ TEST_F(cls_log, trim_by_marker)
   utime_t start_time = ceph_clock_now();
   generate_log(ioctx, oid, 10, start_time, true);
 
-  utime_t zero_time;
```
Yes, thanks for the reminder. I have now created a PR for this.

jmundack replied:
FYI - got the PR merged

phlogistonjohn commented on src/test/fio/fio_librgw.cc:
```diff
@@ -312,9 +310,9 @@ namespace {
 
       if (data->bucket_fh) {
-	r = rgw_fh_rele(data->fs, data->bucket_fh, 0 /* flags */);
+	rgw_fh_rele(data->fs, data->bucket_fh, 0 /* flags */);
```
Yes, thanks for the reminder. I have also created a PR for this.

jmundack replied:
FYI - got the PR merged

jmundack commented:
general question - do we want to add some non blocking (while we are flushing out the details) PR checks to make sure the containerized builds wont break on a code/dependency change?

phlogistonjohn commented:
> general question - do we want to add some non blocking (while we are flushing out the details) PR checks to make sure the containerized builds wont break on a code/dependency change?

I'm open to the idea, but I'm not sure where to begin. What are you thinking of in particular?

phlogistonjohn commented:
I've cleaned up the content of this branch a lot. I've removed a lot of the WIP patches that existed to get certain builds passing - you can still run them, but failures are expected. 
The WIP stuff now exists in my [jjm-containerized-build-wip1 branch](https://github.com/ceph/ceph/compare/main...phlogistonjohn:ceph:jjm-containerized-build-wip1)

This should make the current PR easier to try out and to review, but if you do want to experiment with it in practice you may still end up need patches from that other branch.

djgalloway commented on Dockerfile.build:
```diff
@@ -0,0 +1,32 @@
+ARG CEPH_CTR_SRC=/usr/local/src/ceph
+ARG CLEAN_DNF=yes
+ARG CEPH_BRANCH=main
+COPY --from=bootstrap ${CEPH_CTR_SRC} ${CEPH_CTR_SRC}
```
Is this line accurate?  COPY from /usr/local/src/ceph to /usr/local/src/ceph?

phlogistonjohn replied:
Yes, the `--from=bootstrap` part means copy said path from the build container named `boostrap`. This container (layer) was already created in previous steps and populated `/usr/local/src/ceph`. (Maybe think of it more like scp'ing from host1's /usr/local/src/ceph to host2 of the same dir?)

phlogistonjohn replied:
If you're curious why I did it this way:
* it produces fewer layers - the COPY statements in the bootstrap build container each produce a layer - but this single COPY statement produces one layer in the resulting image
* it felt nice to organize the "minimal" set of files taken from the ceph tree needed to create the build image all together in one logical block

jmundack commented:
@phlogistonjohn  - can we also add some instruction on how to build/test this? (i am hoping to play with it and also help fix some of the open trackers linked here)

phlogistonjohn commented:
new instructions have been added to the PR description under `How to test & try`. I wrote it fast and can't really test it myself because I'm the author :-) So feedback on that is very much welcome.

dparmar18 commented:
I like this initiative @phlogistonjohn. This PR made me recall making use of Rook for testing my NFS related patches last year. It was as simple as having a Dockerfile with custom built binaries, add it to my quay account and use it in config YAML to spin up a cluster. So, having something like this within Ceph repo is really helpful. I'll plan to use this in some of my upcoming testings.

jmundack commented:
one feedback item
- having an option to override the url for the base OS image and what the OS is (centos vs ubuntu) would be useful for environments that don't have connectivity to the internet or i you want to use a custom local image

phlogistonjohn commented:
The latest push adds:
* a --base-image option to override the image used by the distro option requested by @jmundack 
* a --dry-run option that prints but does not run "major" commands requested by @zmc 

For `--base-image` you can use it like `python3 ./src/script/build-with-container.py -d ubuntu22.04 -b build.u24/  -e build  --base-image=registry.exmaple.com/mirrored/ubuntu:22.04` and it should use `registry.example.com/mirrored/ubuntu:22.04` instead of the default of `docker.io/ubuntu:22.04` when creating the build image. Note that creating the build image will attempt to run apt-get/dnf commands, so your base image should either be configured to use your package mirrors if inside a firewall or allow those commands to reach the stock distro package repos.

For both, please do test it a bit as I have only had limited testing of it myself so far.

jmundack commented:
@phlogistonjohn  - https://github.com/ceph/ceph/pull/60986 is the clean up PR to fix most of the trackers you have listed for warnings/errors

deepssin commented:
Hello @phlogistonjohn 

I tried out the build-with-containers.py as mentioned in `test & try` section. Though I wasn't able to check out to your branch[permission issues].

I cloned your changes into my local and tried using command - `python3 src/script/build-with-containers.py -d centos9 -e build`

```
+ DISTRO_KIND=centos:9
+ case "${CEPH_BRANCH}~${DISTRO_KIND}" in
+ echo 'Unknown action, branch or build: main~centos:9'
Unknown action, branch or build: main~centos:9
+ exit 2
```
Please let me know if I am missing something here?

phlogistonjohn commented:
> Hello @phlogistonjohn
> 
> I tried out the build-with-containers.py as mentioned in `test & try` section. Though I wasn't able to check out to your branch[permission issues].
> 
> I cloned your changes into my local and tried using command - `python3 src/script/build-with-containers.py -d centos9 -e build`
> 
> ```
> + DISTRO_KIND=centos:9
> + case "${CEPH_BRANCH}~${DISTRO_KIND}" in
> + echo 'Unknown action, branch or build: main~centos:9'
> Unknown action, branch or build: main~centos:9
> + exit 2
> ```
> 
> Please let me know if I am missing something here?

Fixed. Sorry for that, I had to make some changes to the script for the new ` --base-image ` option and then I only tested on ubuntu after those changes were made.

ronen-fr commented:
@phlogistonjohn hi, John. Two question:
1. how to prevent the script from filling up /var/tmp? 
when trying (for example: `./src/script/build-with-container.py -d centos9 -b configure -e tests --homedir ~/src/cont59841 --keep-container`, with or without the homedir param), I get:

_> [2/2] COMMIT ceph-build:ctr1.centos9
> Error: committing container for step {Env:[CEPH_CTR_SRC=/usr/local/src/ceph CLEAN_DNF=yes CEPH_BRANCH=main DISTRO=quay.io/centos/centos:stream9 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin FOR_MAKE_CHECK=1] Command:run Args:[DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh] Flags:[] Attrs:map[] Message:RUN DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh Heredocs:[] Original:RUN DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh}: copying layers and metadata for container "04572749a7b8431ab8e7bed83e3bc295e7c00d3ab41a210d8825384bb8e30211": writing blob: storing blob to file "/var/tmp/container_images_storage108785264/1": write /var/tmp/container_images_storage108785264/1: no space left on device
> 2025-02-04 07:31:33,502: ERROR: Command failed: Command '['podman', 'build', '-t', 'ceph-build:ctr1.centos9', '--build-arg=JENKINS_HOME=/home/rfriedma/src/cont59841', '--build-arg=DISTRO=quay.io/centos/centos:stream9', '--volume=/mnt/nvme0/src/cont59841/ceph:/home/rfriedma/src/cont59841:Z', '-f', 'Dockerfile.build', '.']' returned non-zero exit status 125._

And - 2- how to create the build environment for Reef?
tried '`CEPH_REF=reef ./src/script/build-with-container.py ...`'
but it seems that 'main' was still the one selected.

Thanks!

phlogistonjohn commented:
Thanks for trying it out!

> @phlogistonjohn hi, John. Two question:
> 
>     1. how to prevent the script from filling up /var/tmp?
>        when trying (for example: `./src/script/build-with-container.py -d centos9 -b configure -e tests --homedir ~/src/cont59841 --keep-container`, with or without the homedir param), I get:
> 
> 
> _> [2/2] COMMIT ceph-build:ctr1.centos9
> 
> > Error: committing container for step {Env:[CEPH_CTR_SRC=/usr/local/src/ceph CLEAN_DNF=yes CEPH_BRANCH=main DISTRO=quay.io/centos/centos:stream9 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin FOR_MAKE_CHECK=1] Command:run Args:[DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh] Flags:[] Attrs:map[] Message:RUN DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh Heredocs:[] Original:RUN DISTRO=$DISTRO     CEPH_BRANCH=$CEPH_BRANCH     CLEAN_DNF=$CLEAN_DNF     CEPH_CTR_SRC=${CEPH_CTR_SRC}     bash -x ${CEPH_CTR_SRC}/buildcontainer-setup.sh}: copying layers and metadata for container "04572749a7b8431ab8e7bed83e3bc295e7c00d3ab41a210d8825384bb8e30211": writing blob: storing blob to file "/var/tmp/container_images_storage108785264/1": write /var/tmp/container_images_storage108785264/1: no space left on device
> > 2025-02-04 07:31:33,502: ERROR: Command failed: Command '['podman', 'build', '-t', 'ceph-build:ctr1.centos9', '--build-arg=JENKINS_HOME=/home/rfriedma/src/cont59841', '--build-arg=DISTRO=quay.io/centos/centos:stream9', '--volume=/mnt/nvme0/src/cont59841/ceph:/home/rfriedma/src/cont59841:Z', '-f', 'Dockerfile.build', '.']' returned non-zero exit status 125._
> 


Interesting. I have not seen that before. Are you able to build container images successfully with podman normally? If yes let's try to execute the build command (seen in the output - minus the comma separators) without the script involved.  If no, we will probably need to debug your podman/system setup a bit.



> And - 2- how to create the build environment for Reef? tried '`CEPH_REF=reef ./src/script/build-with-container.py ...`' but it seems that 'main' was still the one selected.
> 

We would need to PR #48697 and these patches to reef. I tried to do a [quick and dirty backport here](https://github.com/ceph/ceph/compare/reef...phlogistonjohn:ceph:jjm-EXP-reef-bwc). I was able to build a ubuntu 22.04 builder image but then ceph failed to configure due to some dpdk error. Unfortunately , I don't have time to debug that today. Maybe later this week.

bill-scales commented:
@ronen-fr we use podman quite a bit and have also had issues with /var/ filling up (and $HOME when not running as root). Other than increasing the size of /var and $HOME there are a couple of solutions:

1. Use the environment variable TMPDIR to have podman place container images somewhere other than /var/tmp
or
2. Create symbolic links for /var/lib/containers and $HOME/.local/share/containers/ to a directory that has more space. If the system has selinux then moving /var/lib/containers can cause problems - you need something like this to keep it happy:

    mv /var/lib/containers/ /extra/containers
    ln -s /extra/containers/ /var/lib/containers
    semanage fcontext -a -e /var/lib/containers /extra/containers
    restorecon -R -v /extra/containers

ronen-fr commented:
@bill-scales 
> @ronen-fr we use podman quite a bit and have also had issues with /var/ filling up (and $HOME when not running as root). Other than increasing the size of /var and $HOME there are a couple of solutions:
> 
Thanks a lot!

zmc commented:
@phlogistonjohn @ronen-fr @bill-scales 

To free up space, I personally use `podman system prune` a lot. A larger hammer is `podman system reset`

zmc commented:
I think this is ready to merge and iterate on. I am going to start work on integrating it with the new build pipeline, which should also be ready to merge: https://github.com/ceph/ceph-build/pull/2310

fayak commented:
Hello @phlogistonjohn !
Thanks for the tool, I could definitely benefit from it :)
Is it planned to work with `docker` ? 
When trying to build something I have
```
2025-02-17 17:22:53,005: INFO: Executing command: docker build -t ceph-build:main.centos9 --build-arg=JENKINS_HOME=/ceph --build-arg=DISTRO=quay.io/centos/centos:stream9 --volume=/home/zarak/Docs/ceph:/ceph:Z -f Dockerfile.build .
unknown flag: --volume
See 'docker buildx build --help'.
2025-02-17 17:22:53,116: ERROR: Command failed: Command '['docker', 'build', '-t', 'ceph-build:main.centos9', '--build-arg=JENKINS_HOME=/ceph', '--build-arg=DISTRO=quay.io/centos/centos:stream9', '--volume=/home/zarak/Docs/ceph:/ceph:Z', '-f', 'Dockerfile.build', '.']' returned non-zero exit status 125.
```

I can see that the `--volume` option is used but it is not supported by docker buildx. Am I missing something like some upcoming update of docker supporting this option, or is it some podman only option ?

phlogistonjohn commented:
It is supposed to work, but I never test it with docker. I rely on others to try it with docker.

Since docker apparently doesn't support passing the --volume option to build, you can try commenting out the section:
```
    if ctx.cli.homedir:
        cwd = pathlib.Path(".").absolute()
        cmd.append(f"--volume={cwd}:{ctx.cli.homedir}:Z")
```
in the `build_container` funciton.

Otherwise please file a tracker issue on ceph: https://tracker.ceph.com - use the `ceph` *project* ( https://tracker.ceph.com/projects/ceph/issues/new ) and the `build` *category*. I will try to look into it soon. Thanks.
